go-controller/pkg/ovn/egressip.go

```diff
@@ -772,8 +772,18 @@ func (oc *Controller) executeCloudPrivateIPConfigChange(egressIPName string, toA
 			}
 		}
 	}
+	// Merge ops into the existing pendingCloudPrivateIPConfigsOps.
+	// This allows us to:
+	// a) execute only the new ops
+	// b) keep track of any pending changes
 	if len(ops) > 0 {
-		oc.eIPC.pendingCloudPrivateIPConfigsOps[egressIPName] = ops
+		if _, ok := oc.eIPC.pendingCloudPrivateIPConfigsOps[egressIPName]; !ok {
+			oc.eIPC.pendingCloudPrivateIPConfigsOps[egressIPName] = ops
+		} else {
+			for cloudPrivateIP, op := range ops {
+				oc.eIPC.pendingCloudPrivateIPConfigsOps[egressIPName][cloudPrivateIP] = op
```

Comment:
@askaris One additional question here, and sorry if irrelevant, I am 100% familiar with what is going on here. If there is already a `toDelete` for that `cloudPrivateIP`, and here we are merging a `toAdd`, shouldn't we keep around the`toDelete` so that it is processed as an update?

So basically the overall merge would be to only replace `toAdd` instead of the whole thing.

Reply:
I only tested the case where an EgressIP's CloudPrivateIP changes. So, egressip "egressip-test" updates its IP address from 192.168.123.10 to 192.168.123.11. That's a concurrent delete for one, and an add for the other. And the code before was prone to that  race condition. 
For your last comment: that would happen if 192.168.123.11 was assigned and the the cache had a delete operation from the existing node, and then an add operation to another node was added on top of it. Or if an add op was being processed, but with a simultaneous delete op. I didn't dare address that case because it's nothing where I observed a bug and I have no means of testing it. I assume though that something like the following might be safe, but I don't know:
~~~
   775     // Merge ops into the existing pendingCloudPrivateIPConfigsOps.
   776     // This allows us to:
   777     // a) execute only the new ops
   778     // b) keep track of any pending changes
   779     if len(ops) > 0 {
   780         if _, ok := oc.eIPC.pendingCloudPrivateIPConfigsOps[egressIPName]; !ok {
   781             // Set all operations for the EgressIP object if none are in the cache currently.
   782             oc.eIPC.pendingCloudPrivateIPConfigsOps[egressIPName] = ops
   783         } else {
   784             for cloudPrivateIP, op := range ops {
   785                 if _, ok := oc.eIPC.pendingCloudPrivateIPConfigsOps[egressIPName][cloudPrivateIP]; !ok ||
   786                     op.toAdd != "" && op.toDelete != "" {
   787                     // a) If this specific EgressIP object's CloudPrivateIPConfig address currently has no
   788                     // op, simply set it.
   789                     // b) Also simply set it if the new op is an update.
   790                     oc.eIPC.pendingCloudPrivateIPConfigsOps[egressIPName][cloudPrivateIP] = op
   791                 } else if op.toAdd != "" {
   792                     // If this EgressIP object's CloudPrivateIPConfig already has an operation, then there are
   793                     // the following possible scenarios:
   794                     // x.x.x.x is about to be added and just received a delete request:
   795                     //  oc.executeCloudPrivateIPConfigOps will expect that the CloudPrivateIPConfig already exist.
   796                     //  If we set this to a delete now, it will become an update operation that might potentially fail
   797                     //  because the CloudPrivateIPConfig does not exist. So ignore this case.
   798                     // x.x.x.x is about to be deleted from a node and now gets an add op, then it's an update and
   799                     //  it should be save to merge in the toAdd op.
   800                     oc.eIPC.pendingCloudPrivateIPConfigsOps[egressIPName][cloudPrivateIP].toAdd = op.toAdd
   801                 }
   802             }
   803         }
   804     }
~~~
Given that I can't really test this, I'm thinking about just fixing the issue that I found, and not changing the current logic for the other cases (= just keeping the override)